### ISSUE TITLE
Ensure Content-Length header is not set on DELETE requests

### ISSUE THREAD

**Issue by andrewbranch** [emberjs/data#3010]
## TL;DR
Interesting issue that took a while to track down: Safari is adding `Content-Length: 0` to HTTP DELETE requests, and Chrome is not. I think that Safari is wrong, and Ember Data can remedy it.

## Why this is a problem
Here is how I discovered this issue:

1. Kick off a DELETE request via `destroyRecord()` or something
2. Hits my express server
3. Request handled by [json-api](https://github.com/ethanresnick/json-api) controller
4. json-api determines whether or not there’s a request body to parse by checking for either a `Transfer-Encoding` or `Content-Length` header
5. Sees a `Content-Length` header so starts figuring out how to parse the body
6. Checks the Content-Type with [content-type](https://github.com/jshttp/content-type)
7. THERE IS NO CONTENT-TYPE DROP THE MIC AND WALK AWAY (server crashes)

I did a lot of research trying to figure out what the expected behavior is, and I’ve linked everything I found below, but the gist is this: having a `Content-Length` header, even if the value is 0, indicates that the request has a body. If the request has a body, the server should interpret the body based on the value of the `Content-Type` header. Right now, DELETE requests (from Safari) have a `Content-Length` header but no `Content-Type`. As such, my server is getting confused.

There are two possible solution:

1. Ensure the `Content-Length` is never set for DELETEs
2. Ensure the `Content-Type` is always set for DELETEs

My vote is option 1, based on my interpretation of the HTTP spec, which isn’t 100% clear on this. Here’s what I found and how I interpret it:

[HTTP/1.1 RFC7231](https://tools.ietf.org/html/rfc7231#section-4.3.5) says
> A payload within a DELETE request message has no defined semantics; sending a payload body on a DELETE request might cause some existing implementations to reject the request.

Pair that with [HTTP/1.1 RFC7230](https://tools.ietf.org/html/rfc7230#section-3.3.2)
> A user agent SHOULD send a Content-Length in a request message when no Transfer-Encoding is sent and the request method defines a meaning for an enclosed payload body.  For example, a Content-Length header field is normally sent in a POST request even when the value is 0 (indicating an empty payload body).  A user agent SHOULD NOT send a Content-Length header field when the request message does not contain a payload body and the method semantics do not anticipate such a body.

So the question that guides us to specification compliance is this: *Is a DELETE request supposed to have an empty body, or no body at all?*

I take all this to mean that DELETEs *shouldn’t* have a body at all, and therefore *shouldn’t* have a `Content-Length` header.

If you disagree and think that DELETEs should have a zero-length body, then I think it follows that they should have a `Content-Type` header (option 2):

> A sender that generates a message containing a payload body SHOULD generate a Content-Type header field in that message unless the intended media type of the enclosed representation is unknown to the sender. [[HTTP/1.1 RFC7231 3.1.1.5](https://tools.ietf.org/html/rfc7231#section-3.1.1.5)]

## Where to go from here
If you’re in agreement that either of the two solutions I proposed should happen, I’m happy to look into it myself, write some failing tests, and submit a PR.

**Comment by igorT** [comment]
Thanks for the very detailed issue. I am not sure what the right approach here is, will raise this at the weekly ED team meeting

**Comment by andrewbranch** [comment]
Thanks @igorT!

A few updates:

1. I’m not sure it’s possible to unset the Content-Length header. Trying to `xhr.setRequestHeader(undefined)` does nothing, and `xhr.setRequestHeader(null)` throws an error about setting an unsafe header.
2. I’m not sure setting Content-Type header with a zero-length header is appropriate either. The server libraries I’m depending on, when seeing a `Content-Type: application/json` header, tries to run `JSON.parse` on an empty string, which throws a parse error. Which kind of makes sense, you know?

So I don’t really have much of an opinion on what I think a good solution would be anymore. I know that if this *can* be considered a bug, it’s really a bug with Safari/WebKit, but I had hoped that Ember Data could do something to make sure requests that it sends are uniformly correct across all browsers.

**Comment by tomdale** [comment]
Hey @andrewbranch,

Thank you for the well-written bug report. We discussed this at the core team meeting and our current thinking is that the correct places to fix this are:

1. In Safari, obviously
2. In the server library consuming the request

@wycats pointed out that Rails has had to fix similar issues when processing malformed Safari requests. Especially if the library crashes when parsing malformed requests, it should be modified to handle that input more gracefully as this is a likely security issue.

If there is an easy fix in Ember Data, we would consider including it, but it doesn't seem likely that we can change the outgoing request after Safari modifies it.

**Comment by andrewbranch** [comment]
@tomdale Thanks for looking into it! I suspected that would be the verdict, but wanted to pass along what I found. I've long since been just deleting the content-length header on incoming DELETE requests in my server code. Oh well.
